File: IFATCCompanion/Weather/TAFParser.swift
import Foundation

/// Parses TAFs from the Aviation Weather Center JSON API (best-effort).
enum TAFParser {

    static func parseJSON(_ data: Data) -> [TAF] {
        JSONLenient.array(data).compactMap { obj in
            guard let icao = JSONLenient.string(obj["icaoId"]) ?? JSONLenient.string(obj["station_id"]) else { return nil }
            var taf = TAF(icao: icao, raw: JSONLenient.string(obj["rawTAF"]) ?? JSONLenient.string(obj["raw_text"]) ?? "")
            taf.issueTime = JSONLenient.date(obj["issueTime"]) ?? JSONLenient.date(obj["bulletinTime"])
            if let fcsts = obj["fcsts"] as? [[String: Any]] {
                taf.periods = fcsts.map { f in
                    TAFForecastPeriod(
                        raw: JSONLenient.string(f["rawTAF"]) ?? "",
                        windDirection: JSONLenient.int(f["wdir"]),
                        windSpeed: JSONLenient.int(f["wspd"]),
                        visibilitySM: JSONLenient.double(f["visib"]),
                        changeIndicator: JSONLenient.string(f["fcstChange"])
                    )
                }
            }
            return taf
        }
    }
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/weather/TafParser.kt
package com.h3consultingpartners.ifatccompanion.core.weather

/**
 * Parses TAFs from the Aviation Weather Center JSON API (best-effort).
 *
 * Ported from `IFATCCompanion/Weather/TAFParser.swift`.
 */
object TafParser {

    fun parseJson(data: ByteArray): List<TAF> = JsonLenient.array(data).mapNotNull { obj ->
        val icao = JsonLenient.string(obj["icaoId"]) ?: JsonLenient.string(obj["station_id"])
            ?: return@mapNotNull null
        val taf = TAF(
            icao = icao,
            raw = JsonLenient.string(obj["rawTAF"]) ?: JsonLenient.string(obj["raw_text"]) ?: "",
        )
        taf.issueTimeMillis =
            JsonLenient.date(obj["issueTime"]) ?: JsonLenient.date(obj["bulletinTime"])
        JsonLenient.objectArray(obj["fcsts"])?.let { fcsts ->
            taf.periods = fcsts.map { f ->
                TAFForecastPeriod(
                    raw = JsonLenient.string(f["rawTAF"]) ?: "",
                    windDirection = JsonLenient.int(f["wdir"]),
                    windSpeed = JsonLenient.int(f["wspd"]),
                    visibilitySM = JsonLenient.double(f["visib"]),
                    changeIndicator = JsonLenient.string(f["fcstChange"]),
                )
            }
        }
        taf
    }
}
